### PR TITLE
build: fix static build by preventing symbol omissions

### LINKED ISSUE #11199: new registry doesn't work with static builds
## Rationale

Static builds (i.e. --enable-shared=no --enable-static=yes in libmesh config) for moose-based apps don't work right with the new registry.  They compile, but the linker omits symbols from registered moose objects because they aren't referenced anywhere from the main app

## Description

Try to build and run any moose-based app (especially outside the moose repo) and then try to run any input file.  You will get unregistered object errors.

## Impact

Re-enable static binary compilation and distribution of moose apps.  This went un-noticed because static compilation is rarely-used functionality.

### PR DESCRIPTION
**Pull request by rwcarlsen**
By explicitly forcing linker to include symbols for libraries that
aren't referenced in the binary. Because the global registry enables
them to be created dynamically at runtime, objects in modules, etc. are
not ever referenced even though they are still actually used.
Conditionally add special linker flags to force their inclusion in
static builds.

Fixes #11199

### CODE REVIEW

File: framework/app.mk [hunk 1]
```diff
@@ -305,10 +306,40 @@ $(app_test_LIB): $(app_HEADER) $(app_plugin_deps) $(depend_libs) $(app_test_obje
...
+applibs += $(app_test_LIB)
+ifneq ($(libmesh_shared),yes)
+  ifneq (,$(findstring clang,$(CXX)))
+    # replace .a with .dylib for testing in dynamic configuration:
+    applibs := $(foreach lib,$(patsubst %.la,%.a,$(applibs)),-Wl,-force_load,$(lib))
+  else
+  ifneq (,$(findstring g++,$(CXX)))
+    applibs := $(foreach lib,$(patsubst %.la,%.a,$(applibs)),-Wl,--whole-archive,$(lib),--no-whole-archive)
+  else
+  ifneq (,$(findstring mpicxx,$(CXX)))
```
Comment (friedmud) [inline framework/app.mk]: It seems to me that you should find clang/gcc (and what about icc) outside of this logic and then do the if tests so that you don't need this duplication for the mpicxx case.

File: framework/contrib/hit/Makefile [hunk 2]
```diff
@@ -5,7 +5,7 @@ PYTHONCFLAGS ?= `python-config --cflags`
 PYTHONLDFLAGS ?= `python-config --ldflags`
 
 hit: main.cc parse.cc lex.cc lex.h parse.h
-	$(CXX) -std=c++11 -g $(CXXFLAGS) $< parse.cc lex.cc -o $@
+	$(CXX) -std=c++11 -l:libtcmalloc.so.4 -l:libprofiler.so.0 -g $(CXXFLAGS) $< parse.cc lex.cc -o $@
```
Comment (friedmud) [inline framework/contrib/hit/Makefile]: What is this?  It looks unrelated to this PR...

**Review by friedmud — CHANGES_REQUESTED** [review]

File: framework/app.mk (hunk 1 above)
Comment (rwcarlsen) [reply framework/app.mk]: Yeah - I'll take a swing at it.

**Review by permcody — APPROVED** [review]

**Review by friedmud — APPROVED** [review]
